### PR TITLE
Test features of Cluster class 

### PR DESCRIPTION
**Pull request by aya9aladdin**
Changes made in this Pull Request:
 

- I added four new test methods for testing the following Cluster class features:

1- empty Cluster object instantiation
2- error raised when centroid is not in the element list
3- error raised when metadata size is not the same as elements list length
4- cluster iteration is iterating over the elements list
5-  len() return the size of the elements list
6- repr() return the expected string 

- Updated test_cluster_add_metadata() method to check for the error raised when trying to add metadata with length not the same as the cluster size
- Updated spacing typos in two error messages in the Cluster class

PR Checklist
------------
 - [ ] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [ ] Issue raised/referenced?

### CODE REVIEW

File: testsuite/MDAnalysisTests/analysis/test_encore.py
Comment (hmacdope) [inline testsuite/MDAnalysisTests/analysis/test_encore.py]: nitpick but would you be able to give this a slightly more descriptive name? Perhaps `empty_cluster`

File: testsuite/MDAnalysisTests/analysis/test_encore.py
Comment (tylerjereddy) [inline testsuite/MDAnalysisTests/analysis/test_encore.py]: It might make sense to break this down into separate assertions on different lines to keep it readable and to keep the feedback on failure a bit more manageable?

I wonder if we need a new fixture for a single usage of `encore.Cluster()`?

File: testsuite/MDAnalysisTests/analysis/test_encore.py
Comment (hmacdope) [inline testsuite/MDAnalysisTests/analysis/test_encore.py]: Just for clarity would you be able to split this into 4 seperate assertions? I know it takes up more space, but it is much easier to see which one actually failed and easier to read with 4 seperate ones.

**Review by hmacdope — CHANGES_REQUESTED** [review]
Hi just a few changes but some good progress!

File: testsuite/MDAnalysisTests/analysis/test_encore.py
Comment (tylerjereddy) [reply testsuite/MDAnalysisTests/analysis/test_encore.py]: I should note, you may just be copying the style used throughout this file with super-long assertion lines, but it likely needs updating/modernizing in any case (in a separate PR).

Comment (hmacdope) [reply testsuite/MDAnalysisTests/analysis/test_encore.py]: Sorry @tylerjereddy must have reviewed at exactly the same time.

File: testsuite/MDAnalysisTests/analysis/test_encore.py
Comment (aya9aladdin) [reply testsuite/MDAnalysisTests/analysis/test_encore.py]: I'll update it, no problem

File: testsuite/MDAnalysisTests/analysis/test_encore.py
Comment (aya9aladdin) [reply testsuite/MDAnalysisTests/analysis/test_encore.py]: to make sure that I get your thoughts, 
1- I should delete the "empty" fixture cause it's sort of redundant 
2- I should break the assert line into 4 separate ones
3- make a new pull request after committing those changes

Comment (hmacdope) [reply testsuite/MDAnalysisTests/analysis/test_encore.py]: You can keep the empty one, just give it a more instructive name. 

Break the asserts into 4 in the same test

Don't make a new PR, just update by pushing new commits to this PR. :)

File: testsuite/MDAnalysisTests/analysis/test_encore.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_encore.py]: Can you match on some of the text being emitted with the LookupError here please? That can be very useful in checking that you're triggering the right error with a test.

File: testsuite/MDAnalysisTests/analysis/test_encore.py
Comment (aya9aladdin) [reply testsuite/MDAnalysisTests/analysis/test_encore.py]: done ✔️

File: testsuite/MDAnalysisTests/analysis/test_encore.py
Comment (aya9aladdin) [reply testsuite/MDAnalysisTests/analysis/test_encore.py]: That's a good point that I didn't cover, I'm working on it

Comment (aya9aladdin) [reply testsuite/MDAnalysisTests/analysis/test_encore.py]: done ✔️

File: testsuite/MDAnalysisTests/analysis/test_encore.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_encore.py]: Agreed with @tylerjereddy (https://github.com/MDAnalysis/mdanalysis/pull/3566#discussion_r830414316), fixture for a single test isn't worth it.

File: testsuite/MDAnalysisTests/analysis/test_encore.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_encore.py]: Please address PEP8 issues: https://github.com/MDAnalysis/mdanalysis/pull/3566#issuecomment-1072831586

Specifically on this line there should be no white spaces around the operator:
```suggestion
        with pytest.raises(LookupError, match=error_message):
```

**Review by IAlibay — CHANGES_REQUESTED** [review]

File: testsuite/MDAnalysisTests/analysis/test_encore.py
Comment (aya9aladdin) [reply testsuite/MDAnalysisTests/analysis/test_encore.py]: done ✔️

**Comment by aya9aladdin** [comment]
Hi everyone,
shall I make  further modifications for make this PR qualified for merging or it's good so far? @IAlibay @tylerjereddy @hmacdope

**Comment by hmacdope** [comment]
@aya9aladdin would you be able to resolve the comments you have addressed?  Easier to see whats left to go then.

File: testsuite/MDAnalysisTests/analysis/test_encore.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_encore.py]: I'm surprised that pep8speaks didn't complain here, it should be one empty lines between class methods.
```suggestion
```

File: testsuite/MDAnalysisTests/analysis/test_encore.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_encore.py]:
```suggestion
    def test_empty_cluster(self):
```
I see the intent behind the capitalization here, however it's best to keep to pep8 here if possible: https://peps.python.org/pep-0008/#function-and-variable-names

**Review by IAlibay — CHANGES_REQUESTED** [review]
Just a couple of comments based on style (PEP8 can get some time getting used to so don't let it discourage you, even our core developers have a tendency to forget about it).

Also please don't forget to add yourself to the author lists in AUTHORS and CHANGELOG.

Otherwise looks good to me! There are lots of other areas of that code which need testing, you're welcome to add more tests to cover them here or in another PR if you want.

File: testsuite/MDAnalysisTests/analysis/test_encore.py
Comment (aya9aladdin) [reply testsuite/MDAnalysisTests/analysis/test_encore.py]: Ok I'm working on it, I just didn't add more tests cause It was mentioned that the starter issues are limited, so I didn't want to consume it. I will add a couple of tests to cover the whole Cluster class 🙌

Comment (IAlibay) [reply testsuite/MDAnalysisTests/analysis/test_encore.py]: It's completely up to you. If you want to move towards a non-starter issue that's perfectly fine too. I'm happy to go ahead with merging this as it is.

Comment (aya9aladdin) [reply testsuite/MDAnalysisTests/analysis/test_encore.py]: I have added more tests to cover the whole Cluster class + I fixed some typo errors. I'd be happy if you guided me to any improvements need to be done before merging

File: package/CHANGELOG
Comment (hmacdope) [inline package/CHANGELOG]: I'm not sure what @IAlibay or @tylerjereddy  thinks, but if you could make this a little more succinct that would be great.

While it is good to extensively document, we generally use a single point with a few lines per change. :)

**Review by hmacdope — APPROVED** [review]
Looking pretty good to me! Just a nitpick on CHANGELOG.  I will approve pending @IAlibay and @tylerjereddy so review not blocking.

File: package/CHANGELOG
Comment (IAlibay) [reply package/CHANGELOG]: Agreed, test changes whilst great aren't user facing, in most cases we don't actually add changelog entries since they don't lead to changes that users need to know about.

I would either not include a changelog entry or make it very simple here, something like "Added tests for X and fixed minor error message typos" would be more than sufficient.

File: package/CHANGELOG
Comment (IAlibay) [inline package/CHANGELOG]: Unfortunately this isn't how we add things to the changelog either, entries should be under 2.2.0 (see below). You should add your username under the same line that starts with mine, and then the changelog entry under the relevant section after that.

File: testsuite/MDAnalysisTests/analysis/test_encore.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_encore.py]:
```suggestion
        error_message = ("Size of metadata is not equal to the "
            "number of cluster elmements")
```
Under the style that MDAnalysis follows (PEP8), line continuations should use this implicit continuation format where possible.

File: package/MDAnalysis/analysis/encore/clustering/ClusterCollection.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/encore/clustering/ClusterCollection.py]: This is a typo, should be "elements" please do change 


```suggestion
                            "cluster elements")
```

File: testsuite/MDAnalysisTests/analysis/test_encore.py
Comment (IAlibay) [reply testsuite/MDAnalysisTests/analysis/test_encore.py]: Please do update wording based on proposed typo fix.

File: testsuite/MDAnalysisTests/analysis/test_encore.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_encore.py]: See comment above about using implicit line continuations.

**Review by IAlibay — CHANGES_REQUESTED** [review]
Couple of quick review comments

File: package/CHANGELOG
Comment (aya9aladdin) [reply package/CHANGELOG]: I just removed this whole section from the CHANGELOG for simplicity

File: package/MDAnalysis/analysis/encore/clustering/ClusterCollection.py
Comment (aya9aladdin) [reply package/MDAnalysis/analysis/encore/clustering/ClusterCollection.py]: I fixed it. Sorry for not noticing it at the first place

File: testsuite/MDAnalysisTests/analysis/test_encore.py
Comment (aya9aladdin) [reply testsuite/MDAnalysisTests/analysis/test_encore.py]: done

File: package/MDAnalysis/analysis/encore/clustering/ClusterCollection.py
Comment (IAlibay) [reply package/MDAnalysis/analysis/encore/clustering/ClusterCollection.py]: Completely wasn't your fault, it's just a nice thing to fix whilst we're dealing with that part of the code.

File: testsuite/MDAnalysisTests/analysis/test_encore.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_encore.py]: See: https://github.com/MDAnalysis/mdanalysis/pull/3566#issuecomment-1072831586

PEP8 also looks for things like visual indent in continuations, so you would need to do something like:
```suggestion
        error_message = ("Size of metadata is not equal to the "
                         "number of cluster elements")
```

File: testsuite/MDAnalysisTests/analysis/test_encore.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_encore.py]: As above
```suggestion
        error_message = ('Size of metadata having label "label" is '
                         'not equal to the number of cluster elements')
```

File: package/CHANGELOG
Comment (IAlibay) [inline package/CHANGELOG]: The number of empty lines here shouldn't change, although your handle should go on the list at line 16/18.

I'll fix this whilst I update for merge conflicts.

**Review by IAlibay — CHANGES_REQUESTED** [review]
Couple extra tiny things and we should be good, thanks @aya9aladdin

File: testsuite/MDAnalysisTests/analysis/test_encore.py
Comment (aya9aladdin) [reply testsuite/MDAnalysisTests/analysis/test_encore.py]: It's fine now 🙌

File: testsuite/MDAnalysisTests/analysis/test_encore.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_encore.py]: I think they are still off by a single white space, sorry. (note: you can just merge from these suggestions if you want, it's often a lot faster than going back to an editor to make the changes).

```suggestion
                         "number of cluster elements")
```

File: testsuite/MDAnalysisTests/analysis/test_encore.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_encore.py]: As above, should be indented by one more whitespace
```suggestion
                         'not equal to the number of cluster elements')
```

**Review by IAlibay — CHANGES_REQUESTED** [review]

File: testsuite/MDAnalysisTests/analysis/test_encore.py
Comment (aya9aladdin) [reply testsuite/MDAnalysisTests/analysis/test_encore.py]: I just fixed it now

**Review by IAlibay — APPROVED** [review]
awesome pep8speaks is happy so I'm happy, thanks for the contribution @aya9aladdin !

**Comment by IAlibay** [comment]
Before I merge @aya9aladdin, have you already introduced yourself on the MDAnalysis developer list? We ask for everyone to have done so before we merge their first PR at the moment. I can't seem to find your github handle anywhere, but if you have already introduced yourself, can you just reply to the introduction email with your github handle?

**Comment by aya9aladdin** [comment]
Yes I've introduced myself and I believe I replied with my github username. Maybe there was an error and the reply hasn't been sent, I'll reply again now

**Comment by IAlibay** [comment]
> Yes I've introduced myself and I believe I replied with my github username. Maybe there was an error and the reply hasn't been sent, I'll reply again now

Thanks, it could just be me too, but if you can do that it'd be awesome!

**Comment by IAlibay** [comment]
Got it now, thanks!

**Pull request merged**
